include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,137 @@
+                template __call <SharedAccessor>(value, sdata);
+    }
+
+    void build_histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> data)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t buckets_per_thread = (KeyBucketCount + GroupSize - 1) / GroupSize;
+        uint32_t baseIndex = (glsl::gl_WorkGroupID().x * GroupSize) * data.elementsPerWT;
+
+        for (int i = 0; i < buckets_per_thread; i++) {
+            uint32_t prev_bucket_count = GroupSize * i;
+            sdata.set(prev_bucket_count + tid, 0);
+        }
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        for (int i = 0; i < data.elementsPerWT; i++)
+        {
+            uint32_t prev_element_count = GroupSize * i;
+            int j = baseIndex + prev_element_count + tid;
+            if (j >= data.dataElementCount)
+                break;
+            uint32_t k = key.get(j);
+            if (robust && (k<data.minimum || k>data.maximum) )
+                continue;
+            sdata.atomicAdd(k - data.minimum, (uint32_t) 1);
+        }
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+    }
+
+    void histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(ScratchAccessor) scratch, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> data)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t buckets_per_thread = (KeyBucketCount + GroupSize - 1) / GroupSize;
+
+        build_histogram(key, sdata, data);
+
+        uint32_t histogram_value = sdata.get(tid);
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        uint32_t sum = inclusive_scan(histogram_value, sdata);
+        scratch.atomicAdd(tid, sum);
+
+        for (int i = 1; i < buckets_per_thread; i++)
+        {
+            uint32_t prev_bucket_count = GroupSize * i;
+
+            if (tid == GroupSize - 1) {
+                sdata.atomicAdd(prev_bucket_count, sum);
+            }
+
+            sdata.workgroupExecutionAndMemoryBarrier();
+
+            uint32_t index = prev_bucket_count + tid;
+            sum = inclusive_scan(sdata.get(index), sdata);
+
+            scratch.atomicAdd(prev_bucket_count + tid, sum);
+        }
+    }
+                
+    void scatter(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ValueAccessor) val, NBL_REF_ARG(ScratchAccessor) scratch, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> data)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t buckets_per_thread = (KeyBucketCount + GroupSize - 1) / GroupSize;
+
+        build_histogram(key, sdata, data);
+
+        for (int i = 0; i < buckets_per_thread; i++)
+        {
+            uint32_t prev_bucket_count = GroupSize * i;
+            uint32_t index = prev_bucket_count + tid;
+            uint32_t exclusive_value = scratch.atomicSub(index, sdata.get(index)) - sdata.get(index);
```

Comment:
compiler might not optimize out you fetching `sdata.get(index)` twice, because an atomic operation implies a memory barrier, I'd fetch and keep it in a temporary

Reply:
Done